docs/pallas/tpu/pipelining.md

```diff
@@ -65,384 +76,117 @@ Let's talk about the components of this diagram in more detail:
-```python
-block_spec = pl.BlockSpec((256, 512), x_index_map)
-```
+Memory spaces can also be used to specify scratch arguments to a kernel via the `scratch_shapes` argument on `pallas_call`. Scratch buffers are persistent across kernel iterations and are useful for storing intermediate results such as partial accumulations and reductions. A scratch buffer must reside in `VMEM` or `SMEM`.
 
-The `BlockSpec`s for `y` and `z` will be the same as the one for `x`.
-
-+++ {"id": "noybOKghzjwG"}
-
-### Putting it together
-
-We provide these arguments to `pallas_call` via `grid`, `in_specs` and
-`out_specs` (`in_specs` corresponds to the tuple of positional arguments,
-and `out_specs` corresponds to the output).
+As an example for using multiple manual memory space assignments in a kernel, the following program copies a slice of an HBM buffer `x_hbm_ref` into a scratch VMEM buffer `scratch_vmem_ref` before using it for arithmetic and storing the result into an output VMEM buffer:
 
 ```{code-cell}
-:id: ehKAYAwIojfv
-:outputId: 504bab29-83f3-4e1f-8664-1860ad15b6de
-
-def add_matrices_pipelined(x: jax.Array, y: jax.Array) -> jax.Array:
-  block_spec = pl.BlockSpec((256, 512), lambda i: (i, 0))
-  return pl.pallas_call(
-      add_matrices_kernel,
-      out_shape=x,
-      in_specs=[block_spec, block_spec],
-      out_specs=block_spec,
-      grid=(2,)
-  )(x, y)
-
-add_matrices_pipelined(x, y)
+---
+executionInfo:
+  elapsed: 65
+  status: ok
+  timestamp: 1744908591430
+  user:
+    displayName: Justin Fu
+    userId: '17543197034567316452'
+  user_tz: 420
+id: zcqz1CA_o50a
+---
+def hbm_vmem_kernel(x_hbm_ref, out_vmem_ref, scratch_vmem_ref):
+  pltpu.sync_copy(x_hbm_ref.at[0:1], scratch_vmem_ref)
+  out_vmem_ref[...] = scratch_vmem_ref[...] + 1
+
+x = jax.random.uniform(jax.random.key(0), (8, 128), jnp.float32)
+out = pl.pallas_call(hbm_vmem_kernel,
+  in_specs=[pl.BlockSpec(memory_space=pltpu.TPUMemorySpace.ANY)],
+  out_shape=jax.ShapeDtypeStruct((1, 128), jnp.float32),
+  scratch_shapes=(pltpu.TPUMemorySpace.VMEM(shape=(1, 128), dtype=jnp.float32),)
+)(x)
+
+np.testing.assert_allclose(out, x[0:1] + 1)
 ```
 
-+++ {"id": "rkytgIZYzz4t"}
-
-We've only added a little bit of code to our original function to add
-automatic pipelining but the `BlockSpec`s and `grid` do a lot of heavy
-lifting!
-
-How does it work? Well, the `BlockSpec`s provide enough information to start
-*prefetching* blocks of our input from HBM into VMEM.
-For example, if we are starting iteration `i` of our `grid`, we can pass
-`i + 1` into the `index_map` functions to obtain the blocks needed for the
-next iteration. We can then start an asynchronous copy for those blocks.
-Similarly for outputs, we can wait for the outputs of the previous iteration
-to be copied before starting the copy for the current iteration's outputs.
++++ {"id": "AuLfbB34lAX5"}
 
-+++ {"id": "7Xtz9oMs0ZRL"}
+(pallas_tpu_scalar_prefetch)=
 
-### Parameterizing a pipeline
+### Scalar Prefetch
```

Comment:
This doesn't sound like it's related to pipelining and it's not even mentioned in the previous doc. Should we move it to a different doc?

Reply:
Sounds good, removing. It's already in another doc.